### PR TITLE
When reaping deployments, retry deployment Update

### LINKED ISSUE #21245: e2e flake: Kubectl client Kubectl run deployment [AfterEach] should create a deployment from an image
http://pr-test.k8s.io/21058/kubernetes-pull-build-test-e2e-gce/28423/

```
Expected error:
        <*errors.errorString | 0xc2084c7070>: {
            s: "Error running &{/var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gce/kubernetes/platforms/linux/amd64/kubectl [kubectl --server=https://104.197.182.112 --kubeconfig=/var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gce/.kube/config delete deployment e2e-test-nginx-deployment --namespace=e2e-tests-kubectl-3kztl] []  <nil>  Error from server: deployments.extensions \"e2e-test-nginx-deployment\" cannot be updated: the object has been modified; please apply your changes to the latest version and try again\n [] <nil> 0xc2083695a0 exit status 1 <nil> true [0xc208441148 0xc208441168 0xc208441198] [0xc208441148 0xc208441168 0xc208441198] [0xc208441160 0xc208441190] [0x8796a0 0x8796a0] 0xc2083573e0}:\nCommand stdout:\n\nstderr:\nError from server: deployments.extensions \"e2e-test-nginx-deployment\" cannot be updated: the object has been modified; please apply your changes to the latest version and try again\n\nerror:\nexit status 1\n",
        }
        Error running &{/var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gce/kubernetes/platforms/linux/amd64/kubectl [kubectl --server=https://104.197.182.112 --kubeconfig=/var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gce/.kube/config delete deployment e2e-test-nginx-deployment --namespace=e2e-tests-kubectl-3kztl] []  <nil>  Error from server: deployments.extensions "e2e-test-nginx-deployment" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
         [] <nil> 0xc2083695a0 exit status 1 <nil> true [0xc208441148 0xc208441168 0xc208441198] [0xc208441148 0xc208441168 0xc208441198] [0xc208441160 0xc208441190] [0x8796a0 0x8796a0] 0xc2083573e0}:
        Command stdout:
        
        stderr:
        Error from server: deployments.extensions "e2e-test-nginx-deployment" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
        
        error:
        exit status 1
        
    not to have occurred
```

cc @janetkuo @bgrant0607

### LINKED ISSUE #21257: integration flake: kubectl delete deployment/nginx-deployment
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/21058/kubernetes-pull-test-unit-integration/14383/build-log.txt

```
Successful get deployment nginx-deployment {{.spec.replicas}}: 1
(BI0215 02:36:58.517962   30654 event.go:211] Event(api.ObjectReference{Kind:"Deployment", Namespace:"default", Name:"nginx-deployment", UID:"09b3c8a2-d3d0-11e5-a35b-0242ac110004", APIVersion:"extensions", ResourceVersion:"1240", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set nginx-deployment-r5u5h to 1
I0215 02:36:58.550730   30654 event.go:211] Event(api.ObjectReference{Kind:"ReplicaSet", Namespace:"default", Name:"nginx-deployment-r5u5h", UID:"09b4bda1-d3d0-11e5-a35b-0242ac110004", APIVersion:"extensions", ResourceVersion:"1241", FieldPath:""}): type: 'Normal' reason: 'SuccessfulDelete' Deleted pod: nginx-deployment-r5u5h-k40co
I0215 02:36:58.550842   30654 event.go:211] Event(api.ObjectReference{Kind:"ReplicaSet", Namespace:"default", Name:"nginx-deployment-r5u5h", UID:"09b4bda1-d3d0-11e5-a35b-0242ac110004", APIVersion:"extensions", ResourceVersion:"1241", FieldPath:""}): type: 'Normal' reason: 'SuccessfulDelete' Deleted pod: nginx-deployment-r5u5h-in3wz
Error from server: deployments.extensions "nginx-deployment" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
!!! Error in ./hack/test-cmd.sh:1004
  'kubectl delete deployment/nginx-deployment "${kube_flags[@]}"' exited with status 1
```

It looks that we lack retry in the test.

cc @kubernetes/goog-ux

### PR DESCRIPTION
**Pull request by janetkuo**
Fixes #21245; fixes #21257

cc @bgrant0607 @kargakis @ironcladlou @mqliang @soltysh @kubernetes/kubectl

### CODE REVIEW

**Comment by janetkuo** [comment]
@k8s-bot ok to test

File: pkg/kubectl/stop.go [hunk 1]
```diff
@@ -361,21 +361,17 @@ func (reaper *DeploymentReaper) Stop(namespace, name string, timeout time.Durati
...
-	// https://github.com/kubernetes/kubernetes/issues/20966
-	// Instead deployment should be Paused at this point and not at next TODO.
-	deployment.Spec.Paused = false
-	deployment, err = deployments.Update(deployment)
+	deployment, err := reaper.updateDeploymentWithRetries(namespace, name, func(d *extensions.Deployment) {
+		// set deployment's history and scale to 0
+		// TODO replace with patch when available: https://github.com/kubernetes/kubernetes/issues/20527
+		zero := 0
+		d.Spec.RevisionHistoryLimit = &zero
+		d.Spec.Replicas = 0
```
Comment (mqliang) [inline pkg/kubectl/stop.go]: There is a `IntPtr()` helper functions in pkg/util.go:  https://github.com/kubernetes/kubernetes/blob/master/pkg/util/util.go#L120

File: pkg/kubectl/stop.go [hunk 2]
```diff
@@ -361,21 +361,17 @@ func (reaper *DeploymentReaper) Stop(namespace, name string, timeout time.Durati
...
-	// TODO: un-pausing should not be necessary, remove when this is fixed:
-	// https://github.com/kubernetes/kubernetes/issues/20966
-	// Instead deployment should be Paused at this point and not at next TODO.
-	deployment.Spec.Paused = false
-	deployment, err = deployments.Update(deployment)
+	deployment, err := reaper.updateDeploymentWithRetries(namespace, name, func(d *extensions.Deployment) {
+		// set deployment's history and scale to 0
+		// TODO replace with patch when available: https://github.com/kubernetes/kubernetes/issues/20527
+		zero := 0
+		d.Spec.RevisionHistoryLimit = &zero
```
Comment (mqliang) [inline pkg/kubectl/stop.go]: There is a `IntPtr()` helper functions in pkg/util.go: https://github.com/kubernetes/kubernetes/blob/master/pkg/util/util.go#L120

**Comment by mqliang** [comment]
Just one nit.

**Comment by soltysh** [comment]
LGTM

File: pkg/kubectl/stop.go (hunk 2 above)
Comment (janetkuo) [inline pkg/kubectl/stop.go]: Will fix this in another PR since it's unrelated to the flake.

**Comment by janetkuo** [comment]
@k8s-oncall: feel free to manually merge this.

**Comment by janetkuo** [comment]
@k8s-oncall: feel free to manually merge this since it fixes 2 P0 test flakes.

File: pkg/kubectl/stop.go (hunk 2 above)
Comment (janetkuo) [inline pkg/kubectl/stop.go]: Filed #21435

**Comment by brendandburns** [comment]
Manually merging since this flaked again.

**Pull request merged**